usr/src/cmd/sgs/link_audit/man/dumpbind.man
.\" ident	"@(#)dumpbind.man	1.3	05/06/08 SMI"
.\" Copyright 2005 Sun Microsystems, Inc.  All rights reserved.
.\" Use is subject to license terms.
.\"
.\" CDDL HEADER START
.\"
.\" The contents of this file are subject to the terms of the
.\" Common Development and Distribution License, Version 1.0 only
.\" (the "License").  You may not use this file except in compliance
.\" with the License.
.\"
.\" You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
.\" or http://www.opensolaris.org/os/licensing.
.\" See the License for the specific language governing permissions
.\" and limitations under the License.
.\"
.\" When distributing Covered Code, include this CDDL HEADER in each
.\" file and include the License file at usr/src/OPENSOLARIS.LICENSE.
.\" If applicable, add the following below this CDDL HEADER, with the
.\" fields enclosed by brackets "[]" replaced with your own identifying
.\" information: Portions Copyright [yyyy] [name of copyright owner]
.\"
.\" CDDL HEADER END
.\"
.if n .tr \--
.TH dumpbind 1 "18 Nov 1996"
.SH NAME
dumpbind \- dump the bindings found in a bindings.so.1 output file
.SH SYNOPSIS
.B dumpbind
[
.B -pqsc
]
.I bindings_file
.SH AVAILABILITY
.LP
SUNWosdem
.SH DESCRIPTION
.B dumpbind
parses the output file created by the 
.B bindings.so.1
.I Link-Auditing
library and displays a summary of all of the bindings that are
are recorded in it.
.LP
.B bindings.so.1
is a
.I Link-Auditing
library which monitors the bindings of function calls and stores
the data in a common buffer pointed to by BT_OUTPUT.  Because this
library tracks procedure calls via the
.I la_pltenter()
.I Link-Auditing
interface this library will record which functions are called as part of
the execution of a program.  This differs from the
.B symbindrep
example which performs a static analysis of all of the bindings that are
recorded in a program without actually executing it.  Also note that it
is safe to use the same BT_BUFFER for multiple processes running
concurrently.  This means that a whole suite of applications can be
run and the interfaces that are used as part of that suite are recorded.
.LP
The
.B bindings.so.1
library is enabled by the setting of the following environment variables.
.LP
.TP 10
.B LD_AUDIT
set this to the value of the bindings.so.1
.I Link-Auditing
library.  By setting this environment variable the gathering of
bindings data is enabled.
.TP
.B BT_OUTPUT
set this to the name of the common output buffer which will be maintained
durring the data gathering periode.  If this is not set the
output buffer will default to
.I /tmp/bindings.data.
.TP
.B BT_BINDFROM
a colon seperated list of libraries that are to be traced.  Only calls from
these libraries will be traced.  The default is to trace all calls.
.TP
.B BT_BINDTO
a colon seperated list of libraries that are to be traced.  Only calls to
these libraries will be traced.  The default is to trace all calls.
.SH OPTIONS
.TP 10
.B \-p
display output in parseable format.
.TP
.B \-q
query mutex_locks in bindings buffer and display there current status.  This is
when querying a problem with the bindings buffer.
.TP
.B \-c
clear all mutex_locks currently held in the bindings buffer.  
.TP
.B \-s
set master mutex_lock in bindings buffer.  This will cause any processes
currently using this bindings buffer to hang until the lock is freed.
.TP
.B \-b
print hash bucket usage statistics for the bindings buffer.

.SH EXAMPLES
Here is a sample of gathering data using the
.B bindings.so.1
.I Link-Auditing
library and then the examination of that data with the
.B dumpbind
command.
.LP
.nf
.ft 3
polyslo 131% export BT_OUTPUT=/tmp/bindings.data
polyslo 3563% LD_AUDIT=/opt/SUNWonld/lib/bindings.so.1
polyslo 3564% export LD_AUDIT
polyslo 3565% ls /usr/ccs/bin
admin              gprof.flat.blurb   nrform             strip.orig
ar                 help               prof               symorder
ar.orig            ld                 prs                tsort
as                 ld.orig            prt                unget
cdc                lex                ranlib             unifdef
comb               lorder             regcmp             val
delta              m4                 rmdel              vc
dis                make               sact               what
dump               mcs                sccs               yacc
error              mcs.orig           sccsdiff           yaccpar
get                nceucform          size
gprof              ncform             stanswer
gprof.callg.blurb  nm                 strip
polyslo 3566% nm /usr/lib/libdl.so.1
 
 
/usr/lib/libdl.so.1:
 
[Index]   Value      Size    Type  Bind  Other Shndx   Name
 
...
[27]    |      1788|       8|FUNC |GLOB |0    |7      |_dladdr
[32]    |      1772|       8|FUNC |GLOB |0    |7      |_dlclose
[44]    |      1796|       8|FUNC |GLOB |0    |7      |_dldump
[29]    |      1780|       8|FUNC |GLOB |0    |7      |_dlerror
[26]    |      1732|       8|FUNC |GLOB |0    |7      |_dlinfo
[17]    |      1740|       8|FUNC |LOCL |0    |7      |_dlmap
[24]    |      1748|       8|FUNC |GLOB |0    |7      |_dlmopen
[23]    |      1756|       8|FUNC |GLOB |0    |7      |_dlopen
[48]    |      1764|       8|FUNC |GLOB |0    |7      |_dlsym
[38]    |         0|       0|OBJT |GLOB |0    |ABS    |_edata
[21]    |      1824|       0|OBJT |GLOB |0    |8      |_end
[42]    |      1820|       0|OBJT |GLOB |0    |8      |_etext
[30]    |      1804|       8|FUNC |GLOB |0    |7      |_ld_concurrency
[34]    |      1812|       8|FUNC |GLOB |0    |7      |bind_guard
[47]    |      1820|       4|OBJT |GLOB |0    |8      |dbg_mask
[20]    |         0|       0|FILE |LOCL |0    |ABS    |dl.c
[43]    |      1788|       8|FUNC |WEAK |0    |7      |dladdr
[22]    |      1772|       8|FUNC |WEAK |0    |7      |dlclose
[31]    |      1796|       8|FUNC |WEAK |0    |7      |dldump
[45]    |      1780|       8|FUNC |WEAK |0    |7      |dlerror
...
polyslo 136% unset LD_AUDIT
.ft
.fi
.LP
The unseting of LD_AUDIT has turned off the bindings data gathering
and the output can now safely be examined by dumpbind.
.LP
.nf
.ft3
polyslo 3567% unset LD_AUDIT
polyslo 3568% dumpbind /tmp/bindings.data
                           Bindings Summary Report
 
Library                             Symbol                   Call Count
------------------------------------------------------------------------
/usr/lib/libc.so.1                  textdomain                    2
/usr/lib/libelf.so.1                elf_kind                      1
/usr/lib/libc.so.1                  munmap                        1
/usr/lib/libc.so.1                  strlen                      353
/usr/lib/libc.so.1                  .urem                         3
/usr/lib/libc.so.1                  iswprint                    444
/usr/lib/libc.so.1                  .udiv                        11
/usr/lib/libc.so.1                  time                          1
/usr/lib/libc.so.1                  .mul                        361
/usr/lib/libc.so.1                  closedir                      1
/usr/lib/libc.so.1                  _rw_read_held                36
/usr/lib/libelf.so.1                elf_strptr                   34
/usr/lib/libc.so.1                  lstat64                       1
/usr/lib/libc.so.1                  _realbufend                1203
/usr/lib/libc.so.1                  qsort                         3
/usr/lib/libc.so.1                  _write                       64
/usr/lib/libelf.so.1                _elf32_fsize                  7
/usr/lib/libc.so.1                  close                         1
/usr/lib/libelf.so.1                elf_begin                     1
/usr/lib/libelf.so.1                elf_version                   1
/usr/lib/libc.so.1                  _isatty                       2
/usr/lib/libc.so.1                  mmap                          1
/usr/lib/libelf.so.1                elf_end                       1
/usr/lib/libc.so.1                  printf                      628
/usr/lib/libc.so.1                  calloc                        2
...
/usr/lib/libc.so.1                  __flsbuf                    511
/usr/lib/libc.so.1                  _close                        1
/usr/lib/libc.so.1                  _doprnt                     630
/usr/lib/libc.so.1                  ___errno                      3
/usr/lib/libc.so.1                  .umul                        23
/usr/lib/libc.so.1                  _thr_main                     7
/usr/lib/libc.so.1                  open                          1
/usr/lib/libc.so.1                  _sbrk_unlocked                8
/usr/lib/libc.so.1                  mutex_lock                    1
/usr/lib/libc.so.1                  _setbufend                    2
/usr/lib/libc.so.1                  sprintf                       2
/usr/lib/libc.so.1                  setlocale                     2
/usr/lib/libelf.so.1                elf_nextscn                  15
/usr/lib/ld.so.1                    dlsym                         2
/usr/lib/libelf.so.1                _elf_getscn                  34
/usr/lib/libc.so.1                  _sbrk                         8
/usr/lib/libc.so.1                  _open64                       1
/usr/lib/libc.so.1                  _mutex_lock                  88
/usr/lib/libc.so.1                  exit                          2
/usr/lib/libc.so.1                  ioctl                         1
/usr/lib/libc.so.1                  getenv                       17
/usr/lib/libc.so.1                  _mutex_held                  36
------------------------------------------------------------------------
Symbol Count: 87    Call Count: 8272
 
polyslo 3570% 
.ft
.fi
.SH SEE ALSO
.BR ld.so.1 (1)
.br
.TZ LLM


usr/src/cmd/sgs/link_audit/common/dumpbind.c
/*
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License, Version 1.0 only
 * (the "License").  You may not use this file except in compliance
 * with the License.
 *
 * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
 * or http://www.opensolaris.org/os/licensing.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */
/*
 * Copyright 2004 Sun Microsystems, Inc.  All rights reserved.
 * Use is subject to license terms.
 */

#pragma ident	"@(#)dumpbind.c	1.6	05/06/08 SMI"

#include	<stdlib.h>
#include	<unistd.h>
#include	<sys/types.h>
#include	<sys/stat.h>
#include	<sys/lwp.h>
#include	<fcntl.h>
#include	<stdio.h>
#include	<sys/mman.h>
#include	<synch.h>
#include	<errno.h>

#include	"bindings.h"

void
usage()
{
	(void) printf("usage: dumpbind [-pqsc] <bindings.data>\n");
	(void) printf("\t-p\tdisplay output in parsable format\n");
	(void) printf("\t-q\tquery all mutex_locks in data buffer\n");
	(void) printf("\t-c\tclear all mutex_locks in data buffer\n");
	(void) printf("\t-s\tset all mutex_locks in data buffer\n");
	(void) printf("\t-b\tprint bucket usage statistics\n");
}

/*
 * Returns 1 if lock held - 0 otherwise.
 */
static int
query_lock(lwp_mutex_t *lock) {
	if (_lwp_mutex_trylock(lock) == 0) {
		(void) _lwp_mutex_unlock(lock);
		return (0);
	} else
		return (1);
}

static void
query_buffer_locks(bindhead * bhp)
{
	int	i, bkt_locks_held = 0;

	(void) printf("bh_strlock: ");
	if (query_lock(&bhp->bh_strlock) == 1)
		(void) printf("lock held\n");
	else
		(void) printf("free\n");

	(void) printf("bh_lock: ");
	if (query_lock(&bhp->bh_lock) == 1)
		(void) printf("lock held\n");
	else
		(void) printf("free\n");

	(void) printf("Buckets: %d - locks held:\n", bhp->bh_bktcnt);
	for (i = 0; i < bhp->bh_bktcnt; i++) {
		if (query_lock(&bhp->bh_bkts[i].bb_lock) == 1) {
			bkt_locks_held++;
			(void) printf("\tbkt[%d]: lock held\n", i);
		}
	}
	if (bkt_locks_held == 0)
		(void) printf("\tnone.\n");
	else
		(void) printf("\t[%d bucket(s) locked]\n", bkt_locks_held);
}

static void
clear_buffer_locks(bindhead * bhp)
{
	int	i;

	if (query_lock(&bhp->bh_strlock) == 1) {
		(void) _lwp_mutex_unlock(&bhp->bh_strlock);
		(void) printf("bh_strlock: cleared\n");
	}
	if (query_lock(&bhp->bh_lock) == 1) {
		(void) _lwp_mutex_unlock(&bhp->bh_lock);
		(void) printf("bh_lock: cleared\n");
	}
	for (i = 0; i < bhp->bh_bktcnt; i++) {
		if (query_lock(&bhp->bh_bkts[i].bb_lock) == 1) {
			(void) _lwp_mutex_unlock(&bhp->bh_bkts[i].bb_lock);
			(void) printf("bkt[%d]: lock cleared\n", i);
		}
	}
}

static void
set_buffer_locks(bindhead * bhp)
{
	int	i;

	for (i = 0; i < bhp->bh_bktcnt; i++)
		(void) _lwp_mutex_lock(&bhp->bh_bkts[i].bb_lock);

	(void) _lwp_mutex_lock(&bhp->bh_strlock);
	(void) _lwp_mutex_lock(&bhp->bh_lock);
}

int
main(int argc, char **argv)
{
	int		fd;
	char		*fname, *format_string;
	bindhead	*bhp, *tmp_bhp;
	int		i, c;
	int		bflag = 0, pflag = 0, qflag = 0, cflag = 0, sflag = 0;
	ulong_t		symcount, callcount;

	while ((c = getopt(argc, argv, "bspcq")) != EOF)
		switch (c) {
		case 'b':
			bflag++;
			break;
		case 'p':
			pflag++;
			break;
		case 'q':
			qflag++;
			break;
		case 'c':
			cflag++;
			break;
		case 's':
			sflag++;
			break;
		case '?':
			usage();
			return (1);
		}

	if (optind == argc) {
		usage();
		return (1);
	}
	fname = argv[optind];
	if ((fd = open(fname, O_RDWR)) == -1) {
		(void) fprintf(stderr,
		    "dumpbindings: unable to open file: %s\n", fname);
		perror("open");
		return (1);
	}
	/* LINTED */
	if ((bhp = (bindhead *)mmap(0, sizeof (bindhead),
	    (PROT_READ | PROT_WRITE), MAP_SHARED, fd, 0)) == MAP_FAILED) {
		(void) fprintf(stderr, "dumpbind: mmap failed\n");
		perror("mmap");
		return (1);
	}

	if (qflag) {
		query_buffer_locks(bhp);
		return (0);
	}

	if (cflag) {
		clear_buffer_locks(bhp);
		return (0);
	}
	if (sflag) {
		set_buffer_locks(bhp);
		return (0);
	}

	/* LINTED */
	if ((tmp_bhp = (bindhead *)mmap(0, bhp->bh_size,
	    (PROT_READ | PROT_WRITE), MAP_SHARED, fd, 0)) == MAP_FAILED) {
		(void) fprintf(stderr, "dumpbind: remap: mmap failed\n");
		perror("mmap");
		return (1);
	}
	(void) close(fd);

	(void) munmap((void *)bhp, sizeof (bindhead));
	bhp = tmp_bhp;

	if (pflag)
		format_string = "%s|%s|%8d\n";
	else {
		if (!bflag)
		    (void) printf(
			"                           Bindings Summary Report\n\n"
			"Library                             Symbol"
			"                   Call Count\n"
			"----------------------------------------------"
			"--------------------------\n");
		format_string = "%-35s %-25s %5d\n";
	}

	symcount = 0;
	callcount = 0;
	for (i = 0; i < bhp->bh_bktcnt; i++) {
		int		ent_cnt = 0;
		binding_entry *	bep;
		unsigned int	bep_off = bhp->bh_bkts[i].bb_head;

		while (bep_off) {
			/* LINTED */
			bep = (binding_entry *)((char *)bhp + bep_off);
			if (!bflag) {
				/* LINTED */
				(void) printf(format_string,
				    (char *)bhp + bep->be_lib_name,
				    (char *)bhp + bep->be_sym_name,
				    bep->be_count);
				symcount++;
				callcount += bep->be_count;
			}
			bep_off = bep->be_next;
			ent_cnt++;
		}
		if (bflag)
			(void) printf("bkt[%d] - %d entries\n", i, ent_cnt);
	}

	if (!bflag && !pflag)
		(void) printf(
			"----------------------------------------------"
			"--------------------------\n"
			"Symbol Count: %lu    Call Count: %lu\n\n",
			symcount, callcount);

	return (0);
}
